### PR TITLE
[HTTP] Support apple disk image (*.dmg)

### PR DESCRIPTION
**Pull request by manicmaniac**
Add support for libraries distributed in dmg file.

### CODE REVIEW

**Comment by orta** [comment]
This looks cool @manicmaniac - can we get some tests + CHANGELOG entries please?

File: lib/cocoapods-downloader/http.rb [hunk 1]
```diff
@@ -62,6 +65,8 @@ def type_with_url(url)
           :tbz
         elsif path =~ /.(txz|tar\.xz)$/
           :txz
+        elsif path =~ /.dmg$/
```
Comment (kylef) [inline lib/cocoapods-downloader/http.rb]: This dot should be escaped (yes, I realise the other regex are also wrong, that's a separate problem).

```diff
- elsif path =~ /.dmg$/
+ elsif path =~ /\.dmg$/
```

File: lib/cocoapods-downloader/http.rb [hunk 2]
```diff
@@ -100,6 +107,13 @@ def extract_with_type(full_filename, type = :zip)
           tar! 'xfj', unpack_from, '-C', unpack_to
         when :txz
           tar! 'xf', unpack_from, '-C', unpack_to
+        when :dmg
+          plist_s = hdiutil! %(attach -plist -nobrowse #{unpack_from} -mountrandom #{unpack_to})
+          plist = REXML::Document.new plist_s
+          xpath = '//key[.="mount-point"]/following-sibling::string'
+          mount_point = REXML::XPath.first(plist, xpath).text.shellescape
```
Comment (segiddins) [inline lib/cocoapods-downloader/http.rb]: we shouldn't need to shell escape any arguments, as executables are called with an array of arguments, outside of a shell

**Comment by manicmaniac** [comment]
@orta Okay, I will try it.

@kylef Well, should I fix all these problems in this commit? Or create another PR?

@segiddins Thank you for advise, I will fix it.

**Comment by orta** [comment]
@manicmaniac feel free to make new commits, and push them to this branch, we will help you out.

**Comment by manicmaniac** [comment]
I wrote up some specs and fixed the problems.
Now it passes unit tests except for 2 lint error.

> lib/cocoapods-downloader/http.rb:56:7: C: Perceived complexity for type_with_url is too high. [8/7]
> lib/cocoapods-downloader/http.rb:96:7: C: Method has too many lines. [26/24]

I'm afraid that it needs more drastic refactor to pass those lints...

**Comment by segiddins** [comment]
@manicmaniac so sorry for letting this languish, but if you rebase this, RuboCop should pass again :)

**Comment by segiddins** [comment]
`hdiutil` seems not to be installed on travis?

**Comment by neonichu** [comment]
I think the issue is that the cp-downloader specs aren't run on Travis OS X VMs:
https://travis-ci.org/CocoaPods/cocoapods-downloader/jobs/76123291#L1

**Comment by manicmaniac** [comment]
hmm... there are 3 choices:
- send an email to Travis CI and enable the experimental multi OS feature ( `os: osx` )
- or use a dirty hack ( `language: objective-c` )
- use a mock of `hdiutil` in a unit test

I think running Travis CI on OSX is troublesome.
I will try the last one.

**Comment by kylef** [comment]
If I recall correctly, the multi OS feature was enabled for the CocoaPods organisation a while back.

**Comment by manicmaniac** [comment]
@kylef Still It seems to be running on Linux.
Misconfiguration or misunderstanding?

File: lib/cocoapods-downloader/http.rb [hunk 3]
```diff
@@ -1,4 +1,6 @@
 require 'zlib'
+require 'rexml/document'
```
Comment (segiddins) [inline lib/cocoapods-downloader/http.rb]: please only require this inside the `extract_dmg` method?

File: CHANGELOG.md [hunk 4]
```diff
@@ -17,7 +22,6 @@
 * Don't checkout git commits onto a new branch, just use the detached head.  
   [Samuel Giddins](https://github.com/segiddins)
 
-
```
Comment (segiddins) [inline CHANGELOG.md]: this newline is intentional

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,11 @@
 
 ## Master
 
+##### Enhancements
+
++ Support for apple disk image (\*.dmg)
```
Comment (segiddins) [inline CHANGELOG.md]: missing two trailing spaces

**Comment by segiddins** [comment]
@kylef can you review this please? Somehow it slipped by us

**Comment by kylef** [comment]
The code changes seem fine to me, providing we want to add support for dmg this is :+1:.

**Comment by segiddins** [comment]
Thanks @manicmaniac, sorry for the delay in getting this merged.

**Pull request merged**